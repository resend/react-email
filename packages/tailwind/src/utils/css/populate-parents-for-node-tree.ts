import { type CssNode, type List, type ListItem, walk } from 'css-tree';

// This expands the definition for CssNode so that we can add a parent property to all nodes
declare module 'css-tree' {
  interface CssNodeCommon {
    parent?: CssNode;
    parentListItem?: ListItem<CssNode>;
    parentList?: List<CssNode>;
  }
}

export function populateParentsForNodeTree(node: CssNode): void {
  const parentPath: CssNode[] = [];
  walk(node, {
    enter(
      child: CssNode,
      parentListItem: ListItem<CssNode>,
      parentList: List<CssNode>,
    ) {
      child.parent = parentPath[0];
      child.parentListItem = parentListItem;
      child.parentList = parentList;
      parentPath.unshift(child);
    },
    leave() {
      parentPath.shift();
    },
  });
}
