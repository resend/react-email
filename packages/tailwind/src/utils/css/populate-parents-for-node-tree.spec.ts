import { parse } from 'css-tree';
import { populateParentsForNodeTree } from './populate-parents-for-node-tree';

test('populateParentsForNodeTree', () => {
  const ast = parse(`
    .foo { color: red; }
    .bar { background: blue; }

    @media (min-width: 600px) {
      .baz { margin: 10px; }
    }

    .qux:hover { 
      @media (hover: hover) {
        padding: 5px; 
      }
    }
  `);

  populateParentsForNodeTree(ast);

  expect(ast).toMatchSnapshot();
});
