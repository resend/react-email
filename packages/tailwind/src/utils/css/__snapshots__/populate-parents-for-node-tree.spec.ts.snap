// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`populateParentsForNodeTree 1`] = `
{
  "children": [
    {
      "block": {
        "children": [
          {
            "important": false,
            "loc": null,
            "parent": [Circular],
            "property": "color",
            "type": "Declaration",
            "value": {
              "children": [
                {
                  "loc": null,
                  "name": "red",
                  "parent": [Circular],
                  "type": "Identifier",
                },
              ],
              "loc": null,
              "parent": [Circular],
              "type": "Value",
            },
          },
        ],
        "loc": null,
        "parent": [Circular],
        "type": "Block",
      },
      "loc": null,
      "parent": [Circular],
      "prelude": {
        "children": [
          {
            "children": [
              {
                "loc": null,
                "name": "foo",
                "parent": [Circular],
                "type": "ClassSelector",
              },
            ],
            "loc": null,
            "parent": [Circular],
            "type": "Selector",
          },
        ],
        "loc": null,
        "parent": [Circular],
        "type": "SelectorList",
      },
      "type": "Rule",
    },
    {
      "block": {
        "children": [
          {
            "important": false,
            "loc": null,
            "parent": [Circular],
            "property": "background",
            "type": "Declaration",
            "value": {
              "children": [
                {
                  "loc": null,
                  "name": "blue",
                  "parent": [Circular],
                  "type": "Identifier",
                },
              ],
              "loc": null,
              "parent": [Circular],
              "type": "Value",
            },
          },
        ],
        "loc": null,
        "parent": [Circular],
        "type": "Block",
      },
      "loc": null,
      "parent": [Circular],
      "prelude": {
        "children": [
          {
            "children": [
              {
                "loc": null,
                "name": "bar",
                "parent": [Circular],
                "type": "ClassSelector",
              },
            ],
            "loc": null,
            "parent": [Circular],
            "type": "Selector",
          },
        ],
        "loc": null,
        "parent": [Circular],
        "type": "SelectorList",
      },
      "type": "Rule",
    },
    {
      "block": {
        "children": [
          {
            "block": {
              "children": [
                {
                  "important": false,
                  "loc": null,
                  "parent": [Circular],
                  "property": "margin",
                  "type": "Declaration",
                  "value": {
                    "children": [
                      {
                        "loc": null,
                        "parent": [Circular],
                        "type": "Dimension",
                        "unit": "px",
                        "value": "10",
                      },
                    ],
                    "loc": null,
                    "parent": [Circular],
                    "type": "Value",
                  },
                },
              ],
              "loc": null,
              "parent": [Circular],
              "type": "Block",
            },
            "loc": null,
            "parent": [Circular],
            "prelude": {
              "children": [
                {
                  "children": [
                    {
                      "loc": null,
                      "name": "baz",
                      "parent": [Circular],
                      "type": "ClassSelector",
                    },
                  ],
                  "loc": null,
                  "parent": [Circular],
                  "type": "Selector",
                },
              ],
              "loc": null,
              "parent": [Circular],
              "type": "SelectorList",
            },
            "type": "Rule",
          },
        ],
        "loc": null,
        "parent": [Circular],
        "type": "Block",
      },
      "loc": null,
      "name": "media",
      "parent": [Circular],
      "prelude": {
        "children": [
          {
            "children": [
              {
                "condition": {
                  "children": [
                    {
                      "kind": "media",
                      "loc": null,
                      "name": "min-width",
                      "parent": [Circular],
                      "type": "Feature",
                      "value": {
                        "loc": null,
                        "parent": [Circular],
                        "type": "Dimension",
                        "unit": "px",
                        "value": "600",
                      },
                    },
                  ],
                  "kind": "media",
                  "loc": null,
                  "parent": [Circular],
                  "type": "Condition",
                },
                "loc": null,
                "mediaType": null,
                "modifier": null,
                "parent": [Circular],
                "type": "MediaQuery",
              },
            ],
            "loc": null,
            "parent": [Circular],
            "type": "MediaQueryList",
          },
        ],
        "loc": null,
        "parent": [Circular],
        "type": "AtrulePrelude",
      },
      "type": "Atrule",
    },
    {
      "block": {
        "children": [
          {
            "block": {
              "children": [
                {
                  "important": false,
                  "loc": null,
                  "parent": [Circular],
                  "property": "padding",
                  "type": "Declaration",
                  "value": {
                    "children": [
                      {
                        "loc": null,
                        "parent": [Circular],
                        "type": "Dimension",
                        "unit": "px",
                        "value": "5",
                      },
                    ],
                    "loc": null,
                    "parent": [Circular],
                    "type": "Value",
                  },
                },
              ],
              "loc": null,
              "parent": [Circular],
              "type": "Block",
            },
            "loc": null,
            "name": "media",
            "parent": [Circular],
            "prelude": {
              "children": [
                {
                  "children": [
                    {
                      "condition": {
                        "children": [
                          {
                            "kind": "media",
                            "loc": null,
                            "name": "hover",
                            "parent": [Circular],
                            "type": "Feature",
                            "value": {
                              "loc": null,
                              "name": "hover",
                              "parent": [Circular],
                              "type": "Identifier",
                            },
                          },
                        ],
                        "kind": "media",
                        "loc": null,
                        "parent": [Circular],
                        "type": "Condition",
                      },
                      "loc": null,
                      "mediaType": null,
                      "modifier": null,
                      "parent": [Circular],
                      "type": "MediaQuery",
                    },
                  ],
                  "loc": null,
                  "parent": [Circular],
                  "type": "MediaQueryList",
                },
              ],
              "loc": null,
              "parent": [Circular],
              "type": "AtrulePrelude",
            },
            "type": "Atrule",
          },
        ],
        "loc": null,
        "parent": [Circular],
        "type": "Block",
      },
      "loc": null,
      "parent": [Circular],
      "prelude": {
        "children": [
          {
            "children": [
              {
                "loc": null,
                "name": "qux",
                "parent": [Circular],
                "type": "ClassSelector",
              },
              {
                "children": null,
                "loc": null,
                "name": "hover",
                "parent": [Circular],
                "type": "PseudoClassSelector",
              },
            ],
            "loc": null,
            "parent": [Circular],
            "type": "Selector",
          },
        ],
        "loc": null,
        "parent": [Circular],
        "type": "SelectorList",
      },
      "type": "Rule",
    },
  ],
  "loc": null,
  "parent": undefined,
  "type": "StyleSheet",
}
`;
